File: eng-task-grading-react-fe/src/model/attendance-dto.ts
import type { StudentDto } from "./student-dto";

export interface AttendanceDayDto {
  id: number;
  title: string;
  selfAssignKey?: string;
}

export interface AttendanceDayCreateDto {
  title: string;
}

export interface AttendanceDayUpdateDto {
  title: string;
}

export interface AttendanceDto {
  id: number;
  courseId: number;
  title: string;
  minWeight: number | null;
  days: AttendanceDayDto[];
}

export interface AttendanceCreateDto {
  title: string;
  minWeight: number | null;
}

export interface AttendanceUpdateDto {
  title: string;
  minWeight: number | null;
}

export interface AttendanceRecordDto {
  id?: number;
  studentId: number;
  attendanceDayId: number;
  attendanceValueId: number;
}

export interface AttendanceValueDto {
  id: number;
  code: string;
  weight: number;
  title: string;
}

export interface AttendanceSetItemDto {
  attendanceId: number;
  studentId: number;
  value: number;
}

export interface AttendanceSetDto {
  attendances: AttendanceDto[];
  students: StudentDto[];
  items: AttendanceSetItemDto[];
}

export interface NewAttendanceSetForCourseDto {
  attendance: AttendanceDto;
  students: NewAttendanceSetForStudentDto[];
}

export interface NewAttendanceSetForStudentDto {
  studentId: number;
  attendanceDays: AttendanceSetItemDto[];
}

export interface AttendanceDaySetDto {
  students: StudentDto[];
  attendanceDays: AttendanceDayDto[];
  items: AttendanceDaySetRecordDto[];
}

export interface AttendanceDaySetRecordDto {
  id?: number;
  studentId: number;
  attendanceDayId: number;
  attendanceValueTitle: string;
  attendanceValueWeight: number;
}

export interface AttendanceDaySelfSignCreateDto {
  studyNumber: string;
  key: string;
}

export interface AttendanceDaySelfSignSetDto {
  attendanceDayId: number;
  key: string | null;
  selfSigns: AttendanceDaySelfSignDto[];
}

export interface AttendanceDaySelfSignDto {
  id: number;
  student: StudentDto;
  creationDateTime: Date;
  ip: string;
}

File: EngTaskGradingNetBESolution/EngTaskGradingNetBE/Models/Dtos/AttendanceDto.cs
using EngTaskGradingNetBE.Models.DbModel;

namespace EngTaskGradingNetBE.Models.Dtos;

public record AttendanceValueDto(
  int Id,
  string Title,
  double Weight
);

public record AttendanceRecordDto(
  int Id,
  int StudentId,
  int AttendanceDayId,
  int AttendanceValueId
);

public record AttendanceDayDto(
  int Id,
  string Title,
  string? SelfAssignKey
);

public record AttendanceDto(
  int Id,
  string Title,
  double? MinWeight,
  ICollection<AttendanceDayDto> Days
);


public record AttendanceDayCreateDto(
  int AttendanceId,
  string Title
);

public record AttendanceDayUpdateDto(
  string Title
);

public record AttendanceCreateDto(
  string Title,
  double? MinWeight
);

public record AttendanceOverviewDto(
  int AttendanceId,
  Dictionary<StudentDto, double> StudentResults);

public record AttendanceSetItemDto(int AttendanceId, int StudentId, double Value);

public record AttendanceSetDto(
  List<AttendanceDto> Attendances,
  List<StudentDto> Students,
  List<AttendanceSetItemDto> Items);

public record AttendanceDaySetDto(
  List<StudentDto> Students,
  List<AttendanceDayDto> AttendanceDays,
  List<AttendanceDaySetRecordDto> Items
);

public record AttendanceDaySetRecordDto(
  int? Id,
  int StudentId,
  int AttendanceDayId,
  string AttendanceValueTitle,
  double AttendanceValueWeight
);

public record AttendanceDaySelfSignCreateDto(
  string StudyNumber,
  string Key
  );

public record AttendanceDaySelfSignSetDto(
  int AttendanceDayId,
  string? Key,
  List<AttendanceDaySelfSignDto> SelfSigns
  );

public record AttendanceDaySelfSignDto(int Id, StudentDto Student, DateTime CreationDateTime, string IP);

